### PR TITLE
Session/ui

### PR DESCRIPTION
**Pull request by jchapmanio**
#### Description
- `ScreenUIController` split apart into more focused sub-controllers.

- Front end UI additions:
    - Session Player name / Quick join / Browse front screen, `SessionScreenController`
    - Lobby screen for selecting deployment (uses fake data), `LobbyScreenController`
    - Results screen for post-match information (uses fake data), `ResultsScreenController`

- In-game UI additions:
    - a game timer (currently uses Time.time) `GameUITimer` (disabled in prefab by default)
    - kill notifications (needs hooking up to trigger) `KillNotification`, `KillNotificationProvider` (disabled in prefab by default)

- Additional misc classes:
    - `ConnectionStatusUIController` - Used on SessionScreen and LobbyScreen to display connection status in a nice way.
    - `DefaultConnectionAnimController` - Used to control connection button animation in existing default connection screen (Used to be handled in ConnectionController - see next section)
    - `Table`, `TableEntry` - Utilities used to populate Lobby and Results screen tables.
    - `PlayerNameInputController` - Used to sanitise player name entry (pretty bare bones at the moment)
    - `RandomNameGenerator` - Used by test classes to get a random name

- Connection status is now contained within static `ConnectionStateReporter` class, and UI elements that need to know the connection status get their information from this class. `ClientWorkerHandler` and `ConnectionController` have been updated to inform `ConnectionStateReporter` about state changes - this helps keep all connection-related stuff accessible in one location but also helps for learning purposes as people can easily see where the connection status is changed and how. I expect my changes to the existing connection classes will be overwritten which is probably for the best as I didn't -really- know what I was doing with them! (but it does work)

- Moved a number of UI scripts into UI folders.

NB to use new session-based UI flow, the `ClientWorkerHandler` component in the scene must have "Use Session Based Flow" flagged, or else the original default connection flow is used.

#### Tests
These test components were used to confirm behaviour of screens and in-game:
    - `ConnectionStateTester`: Used on a GO with a text component, simply reports the current connection state + extra information (if available)
    - `LobbyScreenTester`: Populates the Lobby screen Table with a defined number of 'deployments' using inspector.
    - `ResultsScreenTester`: Populates the Results screen Table with a defined number of player results using inspector.
    - `KillNotificationProviderTester`: Can be used on KillNotificationProvider GO to spawn random kill notifications using inspector.
    - `SessionFlowTester`: Fakes the gathering of deployments. Used for testing connection status flow when first playing the game.

#### Documentation
Connection flow doc 
https://docs.google.com/drawings/d/1QHrvO_Un20LRpPc_anO1qSERHShM2vG4_rSX9kbXmHw/edit

### CODE REVIEW

File: workers/unity/Assets/Fps/Scripts/SetupLogic/ConnectionStateReporter.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            Connecting,
+            Connected,
+            ConnectionFailed,
+            Spawning,
+            Spawned,
+            SpawningFailed,
+            WorkerDisconnected
+        }
+
+        public static bool AreConnected { get; private set; }
```
Comment (jessicafalk) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/ConnectionStateReporter.cs]: `IsConnected`? or do we actually mean multiple connections?

File: workers/unity/Assets/Fps/Scripts/SetupLogic/ConnectionStateReporter.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            if (!AreConnected)
+            {
+                Debug.LogWarning("Tried spawning when we weren't connected");
+                return;
+            }
+
+            if (AreSpawned)
+            {
+                Debug.LogWarning("Tried spawning when we were already spawned");
```
Comment (jessicafalk) [inline workers/unity/Assets/Fps/Scripts/SetupLogic/ConnectionStateReporter.cs]: who is we? ;)

File: workers/unity/Assets/Fps/Scripts/SetupLogic/ConnectionStateReporter.cs (hunk 1 above)
Comment (jchapmanio) [reply workers/unity/Assets/Fps/Scripts/SetupLogic/ConnectionStateReporter.cs]: "We" are connected, of course :P "we" being the computer and the user...

File: workers/unity/Assets/Fps/Scripts/UI/Hud/KillNotifications/KillNotification.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                yield return null;
+            }
+
+            animator.SetFloat("LifeRemaining", lifeRemaining);
+        }
+
+        // Destroy() is called from animation event
+        private void Destroy()
+        {
+            Destroy(gameObject);
```
[the remark this answers is not in the document]
Comment (jchapmanio) [reply workers/unity/Assets/Fps/Scripts/UI/Hud/KillNotifications/KillNotification.cs]: It does...but it looks confusing doesn't it. `Destroy(gameObject)` is actually `UnityEngine.Object.Destroy()`

[the remark this answers is not in the document]
Comment (jchapmanio) [reply workers/unity/Assets/Fps/Scripts/UI/Hud/KillNotifications/KillNotification.cs]: It seems to.. `Destroy(gameObject)` is actually `UnityEngine.Object.Destroy(gameObject)`. I could rename `Destroy()` to something else?

[the remark this answers is not in the document]
Comment (jchapmanio) [reply workers/unity/Assets/Fps/Scripts/UI/Hud/KillNotifications/KillNotification.cs]: and `OnDestroy()` is the name of the Monobehaviour method that gets auto-called

File: workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                }
+
+                entryHeightCached = EntryTemplate.GetComponent<RectTransform>().rect.height;
+                return entryHeightCached;
+            }
+        }
+
+        private float entryHeightCached = -1f;
+
+        [NonSerialized] private readonly List<TableEntry> Entries = new List<TableEntry>();
```
Comment (jchapmanio) [inline workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs]: I'm not sure there's any benefit to having [NonSerialized] flag on a private List?

File: workers/unity/Assets/Fps/Scripts/UI/Controllers/ResultsScreenController.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        private void AddPlayerToTable(ResultsData data, bool isLocalPlayer)
+        {
+            var entry = (ResultsTableEntry) resultsTable.AddEntry();
+            entry.SetData(data);
+            entry.SetAllTextVisuals(isLocalPlayer ? LocalPlayerTextColor : DefaultTextColor, isLocalPlayer);
+            entry.SetBackgroundColor(isLocalPlayer ? LocalPlayerBackgroundColor : currentBgColor);
+        }
+
+        public class ResultsData
```
[the remark this answers is not in the document]
Comment (jchapmanio) [reply workers/unity/Assets/Fps/Scripts/UI/Controllers/ResultsScreenController.cs]: Should I give it a more descriptive name, then? Any suggestions? `MatchResultsTableData`, `DeploymentTableData`?

File: workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Fps
```
Comment (jchapmanio) [inline workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs]: I think `Fps.UI` is a namespace we use now. I suppose we ought to update all these scripts to be in that namespace?

File: workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs (hunk 4 above)
Comment (jessicafalk) [reply workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs]: probably not

File: workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs (hunk 6 above)
Comment (jessicafalk) [reply workers/unity/Assets/Fps/Scripts/UI/Tables/Table.cs]: yeah, probably need to do a sanity check later on to make sure everything uses the right namespace

**Pull request merged**